### PR TITLE
feat: SpaceRoles in NSTemplateTier

### PR DESCRIPTION
**Pull request by xcoulon**
See also https://github.com/codeready-toolchain/host-operator/pull/591

Fixes https://issues.redhat.com/browse/CRT-1345

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: api/v1alpha1/nstemplatetier_types.go [hunk 1]
```diff
@@ -28,6 +28,11 @@ type NSTemplateTierSpec struct {
 	// +optional
 	ClusterResources *NSTemplateTierClusterResources `json:"clusterResources,omitempty"`
 
+	// the templates to set the spaces roles, indexed by role
+	// +patchStrategy=merge
+	// +mapType=atomic
+	SpaceRoles map[string]NSTemplateTierSpaceRoles `json:"spaceRoles,omitempty" patchStrategy:"merge"`
```
Comment (MatousJobanek) [inline api/v1alpha1/nstemplatetier_types.go:34]: does it really make sense to specify `patchStrategy` to `merge` when the type of the map is supposed to be `atomic`?

In addition, we should mark it as optional (at least for the beginning)

Comment (xcoulon) [reply api/v1alpha1/nstemplatetier_types.go]: ah right, it was supposed to be `optional` but after some changes, I removed that annotation by accident 🤦‍♂️ 
as for the `+patchStrategy=merge` vs `+mapType=atomic`, indeed, it's contradictory. 
See 4c9cd46

note that there are other places where the combination of `+patchStrategy=merge` along with `+mapType=atomic` is used:
- https://github.com/codeready-toolchain/api/blob/master/api/v1alpha1/templateupdaterequest_types.go#L60-L62
- https://github.com/codeready-toolchain/api/blob/master/api/v1alpha1/toolchainstatus_types.go#L99-L101

it's probably worth fixing them, but maybe in a separate PR?

File: api/v1alpha1/nstemplatetier_types.go [hunk 2]
```diff
@@ -45,6 +50,12 @@ type NSTemplateTierClusterResources struct {
 	TemplateRef string `json:"templateRef"`
 }
 
+// NSTemplateTierSpaceRoles the space roles definition in an NSTemplateTier resource
+type NSTemplateTierSpaceRoles struct {
```
Comment (MatousJobanek) [inline api/v1alpha1/nstemplatetier_types.go:54]: btw, the name should have been singular, and not plural 
```suggestion
type NSTemplateTierSpaceRole struct {
```

Comment (xcoulon) [reply api/v1alpha1/nstemplatetier_types.go]: changed in cdea401

**Review by MatousJobanek — APPROVED** [review]

**Pull request merged**